File: packages/dprint-plugin-typescript/src/parsing/nodeHelpers.ts
import * as babel from "@babel/types";
import { BaseContext } from "@dprint/core";
import { BabelToken } from "./BabelToken";

export function hasBody(node: babel.Node) {
    return (node as any as babel.ClassDeclaration).body != null;
}

export function hasSeparatingBlankLine(nodeA: babel.Node | babel.Comment, nodeB: babel.Node | babel.Comment | undefined) {
    if (nodeB == null)
        return false;

    return getNodeBStartLine() > nodeA.loc!.end.line + 1;

    function getNodeBStartLine() {
        const leadingComments = (nodeB! as babel.Node).leadingComments;

        if (leadingComments != null) {
            for (const leadingComment of leadingComments) {
                const commentStartLine = leadingComment.loc!.start.line;
                if (commentStartLine > nodeA.loc!.end.line)
                    return commentStartLine;
            }
        }

        return nodeB!.loc!.start.line;
    }
}

export function getLeadingCommentOnDifferentLine(node: babel.Node, commentsToIgnore?: ReadonlyArray<babel.Comment>) {
    if (node.leadingComments == null)
        return undefined;

    for (const comment of node.leadingComments) {
        if (commentsToIgnore != null && commentsToIgnore.includes(comment))
            continue;

        if (comment.loc!.start.line < node.loc!.start.line)
            return comment;
    }

    return undefined;
}

export function hasLeadingCommentOnDifferentLine(node: babel.Node, commentsToIgnore?: ReadonlyArray<babel.Comment>) {
    return getLeadingCommentOnDifferentLine(node, commentsToIgnore) != null;
}

export function getUseNewlinesForNodes(nodes: ReadonlyArray<babel.Node | BabelToken | null | undefined>) {
    const nonNullNodes = getNodes();
    const firstNode = nonNullNodes.next().value;
    const secondNode = nonNullNodes.next().value;

    if (firstNode == null || secondNode == null || firstNode.loc!.end.line === secondNode.loc!.start.line)
        return false;

    return true;

    function* getNodes() {
        for (const node of nodes) {
            if (node != null)
                yield node;
        }
    }
}

export function isFirstNodeOnLine(node: babel.Node | BabelToken, context: BaseContext) {
    for (let i = node.start! - 1; i >= 0; i--) {
        const char = context.fileText[i];
        if (char === " " || char === "\t")
            continue;

        return char === "\n";
    }

    return true;
}

export function hasParentheses(node: babel.Node): boolean {
    const extra = (node as any).extra;
    if (extra == null)
        return false;
    return extra.parenthesized || false;
}

export function getStartOrParenStart(node: babel.Node): number {
    const extra = (node as any).extra;
    const parenStart = extra && extra.parenStart;
    return parenStart != null ? parenStart : node.start!;
}

export function getJsxText(jsxText: babel.JSXText) {
    // this is necessary because .value will resolve character entities (ex. &nbsp; -> space)
    return (jsxText as any).extra.raw as string;
}

File: packages/rust-dprint-plugin-typescript/src/node_helpers.rs
use super::*;
use swc_common::{comments::{Comment}, BytePos};

pub fn is_first_node_on_line(node: &dyn Ranged, context: &mut Context) -> bool {
    let start = node.lo().0 as usize;

    for i in (0..start).rev() {
        let c = context.file_bytes[i];
        if c != ' ' as u8 && c != '\t' as u8 {
            return c == '\n' as u8;
        }
    }

    return true;
}

pub fn has_separating_blank_line(first_node: &dyn Ranged, second_node: &dyn Ranged, context: &mut Context) -> bool {
    return get_second_start_line(first_node, second_node, context) > first_node.end_line(context) + 1;

    fn get_second_start_line(first_node: &dyn Ranged, second_node: &dyn Ranged, context: &mut Context) -> usize {
        let leading_comments = second_node.leading_comments(context);

        for comment in leading_comments {
            let comment_start_line = comment.start_line(context);
            if comment_start_line > first_node.end_line(context) {
                return comment_start_line;
            }
        }

        second_node.start_line(context)
    }
}

pub fn get_use_new_lines_for_nodes(first_node: &dyn Ranged, second_node: &dyn Ranged, context: &mut Context) -> bool {
    return first_node.end_line(context) != second_node.start_line(context);
}

pub fn has_leading_comment_on_different_line<'a>(node: &dyn Ranged, comments_to_ignore: Option<&Vec<&'a Comment>>, context: &mut Context<'a>) -> bool {
    get_leading_comment_on_different_line(node, comments_to_ignore, context).is_some()
}

pub fn get_leading_comment_on_different_line<'a>(node: &dyn Ranged, comments_to_ignore: Option<&Vec<&'a Comment>>, context: &mut Context<'a>) -> Option<&'a Comment> {
    let comments_to_ignore: Option<Vec<BytePos>> = comments_to_ignore.map(|x| x.iter().map(|c| c.lo()).collect());
    let node_start_line = node.start_line(context);
    for comment in node.leading_comments(context) {
        if let Some(comments_to_ignore) = &comments_to_ignore {
            if comments_to_ignore.contains(&comment.lo()) {
                continue;
            }
        }

        if comment.start_line(context) < node_start_line {
            return Some(comment);
        }
    }

    return None;
}
